### PR TITLE
Added workflow to instantiate VM on EGI

### PR DESCRIPTION
**Pull request by xpivan**
The main purpose of this request is to update a workflow to instantiate a VM. It is based on the jOCCI-API (java Open Cloud Computing Interface) from EGI. 

The workflow works as follow:
The GEF build the docker images based on alpine OS from the internal services "maven-EGI" which contains all the workflow to instantiate the VM. The image is build with the other docker services in pier.go. The workflow starts from the creation of a proxy to send request to EGI server and then read the configEGI.json file located inside the images. At the end of the workflow the code return 3 outputs, the VM IP, URI and state and store it in  configEGI.json.

There is still a lot of work to do:

1) The very first step consists to complete the workflow by automating the configuration of the VM to deploy docker server on EGI based on tcp protocol.
2) Clean and make the java code much more clear
3) The final workflow is supposed to be run inside a docker alpine-java to be lighter. For the moment the docker container will also include maven
4) Structure the json file with more details on users
5) And of course all the suggestions and modifications you will find

### CODE REVIEW

File: gefserver/pier/pier.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ type dockerConnection struct {
 	stageIn        internalImage
 	fileList       internalImage
 	copyFromVolume internalImage
+	mavenEGI	   internalImage
```
Comment (ubuntolog) [inline gefserver/pier/pier.go]: Formatting is wrong. There is "go fmt ./..." command that will take care of tabs and spaces

File: services/_internal/maven-EGI/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,39 @@
+FROM zenika/alpine-maven
+
+# Install bash to allow manual inspection i.e docker run -it $IMAGE_NAME /bin/bash
+RUN apk add --update bash && rm -rf /var/cache/apk/* \
+	apk upgrade && \
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/Dockerfile]: indentation

File: services/_internal/maven-EGI/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# Copy the VOMS project - use to generate the proxy
+COPY VOMS-proxy /home/VOMS-proxy
+
+RUN mvn -f /home/VOMS-proxy/ compile & mvn -f /home/VOMS-proxy/ package
+
+# Copy the maven project inside the image
+COPY workflowEGI /home/workflowEGI
+
+# Compile and package the maven project
+RUN	mvn -f /home/workflowEGI/ compile & mvn -f /home/workflowEGI/ package
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/Dockerfile]: It looks like a tab after RUN

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+					{												
+						if (Attributes[i].contains("occi.networkinterface.address"))
+						{							
+							publicIP = Attributes[i].replace("occi.networkinterface.address=","");
+						}
+				}			System.out.println("[ VM DESCRIPTION - bis ]");
+
+    		for (Entity entity : entities) {
+
+				Map<Attribute, String> map = entity.getAttributes();
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+		} 
+		
+		if (properties.getProperty("RESOURCE").equals("resource_tpl"))
+		{
+			// Getting the description of all the available template(s)
+			System.out.println("[ RESOURCE_TPL DESCRIPTION ]");
+			List<Mixin> mixins = model.findRelatedMixins(properties.getProperty("RESOURCE"));
+
+                        if (!mixins.isEmpty())
+       	        	        for (Mixin mixin : mixins) {
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+	} catch (CommunicationException | RenderingException | AmbiguousIdentifierException ex) {
+		throw new RuntimeException(ex);
+	}
+
+    }
+
+    public static String[] describe (String vmID)
+    {
+    System.out.println("Debut de la description ici...");
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    System.out.println("vmID dans describe: " + vmID);
+	// [ Setting preferences here! ]
+	String AUTH = "x509"; 
+        String OCCI_ENDPOINT_HOST = "https://carach5.ics.muni.cz:11443"; // <= Change here!
+        //String OCCI_ENDPOINT_HOST = "http://stack-server.ct.infn.it:8787/occi1.1"; // <= Change here!
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: Here and below a lot of commented code. If you are not planning to use this code, remove it

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+	             (!RESOURCE.get(i).equals("resource_tpl")) ) 
+        		properties.setProperty("OCCI_RESOURCE_ID", RESOURCE.get(i));
+        		
+		else { 
+			properties.setProperty("RESOURCE", RESOURCE.get(i));
+        		//properties.setProperty("OCCI_RESOURCE_ID", "noempty");
+			}
+        }
+	
+        properties.setProperty("TRUSTED_CERT_REPOSITORY_PATH", TRUSTED_CERT_REPOSITORY_PATH);
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+        properties.setProperty("PROXY_PATH", PROXY_PATH);
+        properties.setProperty("OCCI_AUTH", AUTH);
+
+	try {
+		HTTPAuthentication authentication = new VOMSAuthentication(PROXY_PATH);
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+import cz.cesnet.cloud.occi.parser.MediaType;
+import cz.cesnet.cloud.occi.core.Link;
+import cz.cesnet.cloud.occi.infrastructure.NetworkInterface;
+import cz.cesnet.cloud.occi.infrastructure.IPNetworkInterface;
+import cz.cesnet.cloud.occi.infrastructure.Storage;
+
+import org.apache.commons.codec.binary.Base64;
+
+public class InstantiateVM
+
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java]: empty line?

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+    public static List<String> ATTRIBUTES = Arrays.asList("occi.core.title=EUDAT");
+
+
+    // Instantiation of all EGI input
+    public static String OCCI_ENDPOINT_HOST = null;
+    public static String RES_TPL = null;
+    public static String OS_TPL = null;
+    public static String PUBLIC_KEY_PATH = null;
+    public static String CONTEXT_PATH = null;
+	public static String instantiatedVmId = null;
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java]: indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+    public static List<String> MIXIN = null;
+    public static List<String> CONTEXT = null;
+    
+    public static Boolean verbose = false;
+
+    // Creating a new VM in the OCCI_ENDPOINT_HOST cloud resource
+
+    public static String doCreate (Properties properties, EntityBuilder eb, Model model, Client client)
+    {
+	URI uri_location = null;
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java]: indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/VOMSProxyInit.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+ 	public static void proxyGeneration ()
+	{			
+		String VONAME = "fedcloud.egi.eu"; 
+		String VOMS_PROXY_FILEPATH = "/tmp/x509up_u5040"; 
+		String VOMS_LIFETIME = "12:00";
+		String VOMSES_DIR = "/etc/vomses/";
+		String X509_CERT_DIR = "/etc/grid-security/certificates/";
+		Boolean ENABLE_RFC = true;
+		
+		try {
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/VOMSProxyInit.java]: formatting, not immediately clear when the Try block ends

File: services/_internal/maven-EGI/workflowEGI/src/org/json/HTTPTokener.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+SOFTWARE.
+*/
+
+/**
+ * The HTTPTokener extends the JSONTokener to provide additional methods
+ * for the parsing of HTTP headers.
+ * @author JSON.org
+ * @version 2015-12-09
+ */
+public class HTTPTokener extends JSONTokener {
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/org/json/HTTPTokener.java]: Make the use of { brackets consistent. You have two variants: when { is on the same line with the function or class declaration, or when it is on the next line. Use the first style everywhere

File: services/_internal/maven-EGI/workflowEGI/configEGI_local.json [hunk 15]
```diff
@@ -0,0 +1 @@
+{"proxy":"\/tmp\/x509up_u5040","endpoint":"https:\/\/carach5.ics.muni.cz:11443","resource_tpl":"resource_tpl#extra_large","vmId":"https:\/\/carach5.ics.muni.cz:11443\/compute\/104350","publicIP":"\"147.228.242.226\"","publicKey":"file:\/Users\/pivan\/.ssh\/fedcloud.pub","vmState":"active","os_tpl":"http:\/\/occi.carach5.ics.muni.cz\/occi\/infrastructure\/os_tpl#uuid_07fbd366_bbd2_5ca4_8f47_82c23ebc7d23_warg_default_shared_228","contextualisation":"file:\/Users\/pivan\/.ssh\/EGI-EUDAT-Integration.yml"}
```
Comment (emanueldima) [inline services/_internal/maven-EGI/workflowEGI/configEGI_local.json]: This file should not be part of the PR, as it only makes sense for your setup.

**Comment by ubuntolog** [comment]
General remark: try fix all formatting problems (spaces, tabs, indentation), remove unused (commented) code fragments. Look at my old comments again, you've missed several comments. After we approve the pull request, you will have to pull the most recent master branch (since it has been updated) and rebase you branch.

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+                        List<Entity> entities = client.describe(URI.create(properties.getProperty("OCCI_RESOURCE_ID")));
+
+                        String title = "", size = "", state = "", ID = "", summary = "", allocation = "";
+			String address = "", vlan = "", netid = "", netvlan = "", phydev = "", bridge = "";
+
+                        for (Entity entity : entities) {
+                                ID = entity.getId();
+
+                                Map<Attribute, String> map = entity.getAttributes();
+                                for (Map.Entry<Attribute, String> entry : map.entrySet()) {
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: Code formatting. It is better to make brackets look consistent. Like here you have ")" and "{" on the same line, but most of the time they are on separate lines

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+                                if (allocation != null) System.out.println("occi.network.allocation = " + allocation);
+                                if (address != null) System.out.println("occi.network.address = " + address);
+                                if (netid != null) System.out.println("org.opennebula.network.id = " + netid);
+                                if (netvlan != null) System.out.println("org.opennebula.network.vlan = " + netvlan);
+                                if (phydev != null) System.out.println("org.opennebula.network.phydev = " + phydev);
+                                if (bridge != null) System.out.println("org.opennebula.network.bridge = " + bridge);
+                        }
+                }
+
+		if (properties.getProperty("RESOURCE").equals("os_tpl"))
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: Brackets: not only in this file, but in all files of this pull request

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+	//"https://carach5.ics.muni.cz:11443/storage/3155"); // <= Change here!
+
+	//List<String> RESOURCE = Arrays.asList("network",
+	//"https://carach5.ics.muni.cz:11443/network/6"); // <= Change here!
+
+	List<String> RESOURCE = Arrays.asList("compute",
+	vmID); // <= Change here!
+	
+
+	// INFN-CATANIA-STACK
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: You are exposing a lot of hard-coded URLs and port numbers. Make sure it is not going to cause any security problems. Having hard coded URLs in not a good idea (unless there is no other choice). A single common config file may be a solution

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+		
+		Model model = client.getModel();
+		EntityBuilder eb = new EntityBuilder(model);
+
+		if  (ACTION.equals("describe")) 
+			doDescribe(properties, client, model);
+
+	} catch (CommunicationException ex ) {
+                 throw new RuntimeException(ex);
+        }
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: Indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java [hunk 20]
```diff
@@ -0,0 +1,369 @@
...
+
+	public static String doCreate (Properties properties, EntityBuilder eb, Model model, Client client)
+	{
+	
+	URI uri_location = null;
+	String networkInterfaceLocation = "";
+	String networkInterfaceLocation_stripped = "";
+	Resource vm_resource = null;
+
+	System.out.println();
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java]: Is it needed here?

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java [hunk 21]
```diff
@@ -0,0 +1,369 @@
...
+		
+		Model model = client.getModel();
+                EntityBuilder eb = new EntityBuilder(model);
+
+		instantiatedVmId = doCreate(properties, eb, model, client);
+				
+		return instantiatedVmId;
+	} catch (CommunicationException ex ) {
+                 throw new RuntimeException(ex);
+        }
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/InstantiateVM.java]: Indentation

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/VOMSProxyInit.java [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+        public static boolean isEmpty(String str)
+        {
+		if (str != null && !str.isEmpty()) return false; 
+		else return true; 
+        }
+	
+	/* M	A	I	N */
+ 	public static void proxyGeneration ()
+	{			
+		String VONAME = "fedcloud.egi.eu";
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/VOMSProxyInit.java]: Config file?

File: services/_internal/maven-EGI/VOMS-proxy/src/main/java/it/infn/ct/VOMSProxyInit.java [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+	public static boolean isEmpty(String str)
+	{
+		if str != null && !str.isEmpty() return false; 
+		else return true; 
+	}
+
+	/* M	A	I	N */
+	public static void main (String[] args)
+	{			
+		String VONAME = "fedcloud.egi.eu"; // <= Change here!
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/VOMS-proxy/src/main/java/it/infn/ct/VOMSProxyInit.java]: Hard coded URL

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java [hunk 24]
```diff
@@ -0,0 +1,301 @@
...
+		catch (CommunicationException | RenderingException | AmbiguousIdentifierException ex) 
+		{throw new RuntimeException(ex);}
+	}
+
+	public static String[] describe (String vmID)
+	{
+
+		// [ Setting preferences here! ]
+		String AUTH = "x509"; 
+		String OCCI_ENDPOINT_HOST = "https://carach5.ics.muni.cz:11443";
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/DescribeVM.java]: Hard coded URL

File: services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/configEGI.json [hunk 25]
```diff
@@ -0,0 +1,8 @@
+{
+	"endpoint": "https://carach5.ics.muni.cz:11443",
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/main/java/it/infn/ct/configEGI.json]: It looks like you already have this URL hard coded in multiple files. Why don't you read it from the config file instead? It could make future code modifications much easier

File: services/_internal/maven-EGI/workflowEGI/src/org/json/CookieList.java [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+        for (final Entry<String,?> entry : jo.entrySet()) {
+        	final String key = entry.getKey();
+        	final Object value = entry.getValue();
+            if (!JSONObject.NULL.equals(value)) {
+                if (b) {
+                    sb.append(';');
+                }
+                sb.append(Cookie.escape(key));
+                sb.append("=");
+				sb.append(Cookie.escape(value.toString()));
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/org/json/CookieList.java]: Formatting

File: services/_internal/maven-EGI/workflowEGI/src/org/json/JSONWriter.java [hunk 27]
```diff
@@ -0,0 +1,326 @@
+package org.json;
```
Comment (ubuntolog) [inline services/_internal/maven-EGI/workflowEGI/src/org/json/JSONWriter.java]: You put the entire JSON library in your code. This is why your pull request is so big. Maven should handle all the dependencies. Please remove all those libraries related to JSON (.../src/org/json/....) and all other 3rd party libraries (in case I did not find them) from your pull request. Maven should fetch them, but not you

**Comment by ubuntolog** [comment]
The tests failed
